// astro/dist/core/build/vite-plugin-hoisted-scripts.js
import { viteID } from "../util.js";
import { getPageDataByViteID } from "./internal.js";
function virtualHoistedEntry(id) {
  return id.endsWith(".astro/hoisted.js") || id.endsWith(".md/hoisted.js");
}
function vitePluginHoistedScripts(astroConfig, internals) {
  return {
    name: "@astro/rollup-plugin-astro-hoisted-scripts",
    resolveId(id) {
      if (virtualHoistedEntry(id)) {
        return id;
      }
    },
    load(id) {
      if (virtualHoistedEntry(id)) {
        let code = "";
        for (let path of internals.hoistedScriptIdToHoistedMap.get(id)) {
          code += `import "${path}";`;
        }
        return {
          code
        };
      }
      return void 0;
    },
    async generateBundle(_options, bundle) {
      for (const [id, output] of Object.entries(bundle)) {
        if (output.type === "chunk" && output.facadeModuleId && virtualHoistedEntry(output.facadeModuleId)) {
          const facadeId = output.facadeModuleId;
          const pathname = facadeId.slice(0, facadeId.length - "/hoisted.js".length);
          const vid = viteID(new URL("." + pathname, astroConfig.root));
          const pageInfo = getPageDataByViteID(internals, vid);
          if (pageInfo) {
            pageInfo.hoistedScript = id;
          }
        }
      }
    }
  };
}
export {
  vitePluginHoistedScripts
};
