// astro/dist/core/config.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
import * as colors from "kleur/colors";
import path from "path";
import { pathToFileURL, fileURLToPath } from "url";
import { mergeConfig as mergeViteConfig } from "vite";
import { z } from "zod";
import load from "@proload/core";
import loadTypeScript from "@proload/plugin-tsm";
import postcssrc from "postcss-load-config";
import { arraify, isObject } from "./util.js";
import { appendForwardSlash, trimSlashes } from "./path.js";
load.use([loadTypeScript]);
async function resolvePostcssConfig(inlineOptions, root) {
  if (isObject(inlineOptions)) {
    const options = __spreadValues({}, inlineOptions);
    delete options.plugins;
    return {
      options,
      plugins: inlineOptions.plugins || []
    };
  }
  const searchPath = typeof inlineOptions === "string" ? inlineOptions : fileURLToPath(root);
  try {
    return await postcssrc({}, searchPath);
  } catch (err) {
    if (!/No PostCSS Config found/.test(err.message)) {
      throw err;
    }
    return {
      options: {},
      plugins: []
    };
  }
}
const LEGACY_ASTRO_CONFIG_KEYS = /* @__PURE__ */ new Set([
  "projectRoot",
  "src",
  "pages",
  "public",
  "dist",
  "styleOptions",
  "markdownOptions",
  "buildOptions",
  "devOptions",
  "experimentalIntegrations"
]);
const AstroConfigSchema = z.object({
  adapter: z.object({ name: z.string(), hooks: z.object({}).passthrough().default({}) }).optional(),
  root: z.string().optional().default(".").transform((val) => new URL(val)),
  srcDir: z.string().optional().default("./src").transform((val) => new URL(val)),
  publicDir: z.string().optional().default("./public").transform((val) => new URL(val)),
  outDir: z.string().optional().default("./dist").transform((val) => new URL(val)),
  site: z.string().url().optional().transform((val) => val ? appendForwardSlash(val) : val).refine((val) => !val || new URL(val).pathname.length <= 1, {
    message: '"site" must be a valid URL origin (ex: "https://example.com") but cannot contain a URL path (ex: "https://example.com/blog"). Use "base" to configure your deployed URL path'
  }),
  base: z.string().optional().default("./").transform((val) => val ? appendForwardSlash(trimSlashes(val)) : val),
  trailingSlash: z.union([z.literal("always"), z.literal("never"), z.literal("ignore")]).optional().default("ignore"),
  build: z.object({
    format: z.union([z.literal("file"), z.literal("directory")]).optional().default("directory")
  }).optional().default({}),
  server: z.preprocess((val) => typeof val === "function" ? val({ command: "error" }) : val, z.object({
    host: z.union([z.string(), z.boolean()]).optional().default(false),
    port: z.number().optional().default(3e3)
  }).optional().default({})),
  integrations: z.preprocess((val) => Array.isArray(val) ? val.flat(Infinity).filter(Boolean) : val, z.array(z.object({ name: z.string(), hooks: z.object({}).passthrough().default({}) })).default([])),
  style: z.object({
    postcss: z.object({
      options: z.any(),
      plugins: z.array(z.any())
    }).optional().default({ options: {}, plugins: [] })
  }).optional().default({}),
  markdown: z.object({
    drafts: z.boolean().optional().default(false),
    mode: z.union([z.literal("md"), z.literal("mdx")]).optional().default("mdx"),
    syntaxHighlight: z.union([z.literal("shiki"), z.literal("prism"), z.literal(false)]).optional().default("shiki"),
    shikiConfig: z.any().optional().default({}),
    remarkPlugins: z.array(z.any()).optional().default([]),
    rehypePlugins: z.array(z.any()).optional().default([])
  }).passthrough().optional().default({}),
  vite: z.any().optional().default({}),
  experimental: z.object({
    ssr: z.boolean().optional().default(false),
    integrations: z.boolean().optional().default(false)
  }).optional().default({})
});
async function validateConfig(userConfig, root, cmd) {
  var _a;
  const fileProtocolRoot = pathToFileURL(root + path.sep);
  if (userConfig.hasOwnProperty("renderers")) {
    console.error('Astro "renderers" are now "integrations"!');
    console.error("Update your configuration and install new dependencies:");
    try {
      const rendererKeywords = userConfig.renderers.map((r) => r.replace("@astrojs/renderer-", ""));
      const rendererImports = rendererKeywords.map((r) => `  import ${r} from '@astrojs/${r === "solid" ? "solid-js" : r}';`).join("\n");
      const rendererIntegrations = rendererKeywords.map((r) => `    ${r}(),`).join("\n");
      console.error("");
      console.error(colors.dim("  // astro.config.js"));
      if (rendererImports.length > 0) {
        console.error(colors.green(rendererImports));
      }
      console.error("");
      console.error(colors.dim("  // ..."));
      if (rendererIntegrations.length > 0) {
        console.error(colors.green("  integrations: ["));
        console.error(colors.green(rendererIntegrations));
        console.error(colors.green("  ],"));
      } else {
        console.error(colors.green("  integrations: [],"));
      }
      console.error("");
    } catch (err) {
    }
    process.exit(1);
  }
  let oldConfig = false;
  for (const key of Object.keys(userConfig)) {
    if (LEGACY_ASTRO_CONFIG_KEYS.has(key)) {
      oldConfig = true;
      break;
    }
  }
  if (oldConfig) {
    throw new Error(`Legacy configuration detected. Please update your configuration to the new format!
See https://astro.build/config for more information.`);
  }
  const AstroConfigRelativeSchema = AstroConfigSchema.extend({
    root: z.string().default(".").transform((val) => new URL(appendForwardSlash(val), fileProtocolRoot)),
    srcDir: z.string().default("./src").transform((val) => new URL(appendForwardSlash(val), fileProtocolRoot)),
    publicDir: z.string().default("./public").transform((val) => new URL(appendForwardSlash(val), fileProtocolRoot)),
    outDir: z.string().default("./dist").transform((val) => new URL(appendForwardSlash(val), fileProtocolRoot)),
    server: z.preprocess((val) => typeof val === "function" ? val({ command: cmd === "dev" ? "dev" : "preview" }) : val, z.object({
      host: z.union([z.string(), z.boolean()]).optional().default(false),
      port: z.number().optional().default(3e3)
    }).optional().default({})),
    style: z.object({
      postcss: z.preprocess((val) => resolvePostcssConfig(val, fileProtocolRoot), z.object({
        options: z.any(),
        plugins: z.array(z.any())
      }).optional().default({ options: {}, plugins: [] }))
    }).optional().default({})
  });
  const result = __spreadProps(__spreadValues({}, await AstroConfigRelativeSchema.parseAsync(userConfig)), {
    _ctx: { scripts: [], renderers: [], adapter: void 0 }
  });
  if (!((_a = result.experimental) == null ? void 0 : _a.integrations) && !result.integrations.every((int) => int.name.startsWith("@astrojs/"))) {
    throw new Error([
      `Astro integrations are still experimental.`,
      ``,
      `Only official "@astrojs/*" integrations are currently supported.`,
      `To enable 3rd-party integrations, use the "--experimental-integrations" flag.`,
      `Breaking changes may occur in this API before Astro v1.0 is released.`,
      ``
    ].join("\n"));
  }
  return result;
}
function resolveFlags(flags) {
  return {
    root: typeof flags.root === "string" ? flags.root : void 0,
    site: typeof flags.site === "string" ? flags.site : void 0,
    port: typeof flags.port === "number" ? flags.port : void 0,
    config: typeof flags.config === "string" ? flags.config : void 0,
    host: typeof flags.host === "string" || typeof flags.host === "boolean" ? flags.host : void 0,
    experimentalSsr: typeof flags.experimentalSsr === "boolean" ? flags.experimentalSsr : false,
    experimentalIntegrations: typeof flags.experimentalIntegrations === "boolean" ? flags.experimentalIntegrations : false,
    drafts: typeof flags.drafts === "boolean" ? flags.drafts : false
  };
}
function mergeCLIFlags(astroConfig, flags, cmd) {
  astroConfig.server = astroConfig.server || {};
  astroConfig.experimental = astroConfig.experimental || {};
  astroConfig.markdown = astroConfig.markdown || {};
  if (typeof flags.site === "string")
    astroConfig.site = flags.site;
  if (typeof flags.experimentalSsr === "boolean")
    astroConfig.experimental.ssr = flags.experimentalSsr;
  if (typeof flags.experimentalIntegrations === "boolean")
    astroConfig.experimental.integrations = flags.experimentalIntegrations;
  if (typeof flags.drafts === "boolean")
    astroConfig.markdown.drafts = flags.drafts;
  if (typeof flags.port === "number") {
    astroConfig.server.port = flags.port;
  }
  if (typeof flags.host === "string" || typeof flags.host === "boolean") {
    astroConfig.server.host = flags.host;
  }
  return astroConfig;
}
async function resolveConfigURL(configOptions) {
  const root = configOptions.cwd ? path.resolve(configOptions.cwd) : process.cwd();
  const flags = resolveFlags(configOptions.flags || {});
  let userConfigPath;
  if (flags == null ? void 0 : flags.config) {
    userConfigPath = /^\.*\//.test(flags.config) ? flags.config : `./${flags.config}`;
    userConfigPath = fileURLToPath(new URL(userConfigPath, `file://${root}/`));
  }
  const config = await load("astro", { mustExist: false, cwd: root, filePath: userConfigPath });
  if (config) {
    return pathToFileURL(config.filePath);
  }
}
async function loadConfig(configOptions) {
  const root = configOptions.cwd ? path.resolve(configOptions.cwd) : process.cwd();
  const flags = resolveFlags(configOptions.flags || {});
  let userConfig = {};
  let userConfigPath;
  if (flags == null ? void 0 : flags.config) {
    userConfigPath = /^\.*\//.test(flags.config) ? flags.config : `./${flags.config}`;
    userConfigPath = fileURLToPath(new URL(userConfigPath, pathToFileURL(root)));
  }
  const config = await load("astro", { mustExist: false, cwd: root, filePath: userConfigPath });
  if (config) {
    userConfig = config.value;
  }
  return resolveConfig(userConfig, root, flags, configOptions.cmd);
}
async function resolveConfig(userConfig, root, flags = {}, cmd) {
  const mergedConfig = mergeCLIFlags(userConfig, flags, cmd);
  const validatedConfig = await validateConfig(mergedConfig, root, cmd);
  return validatedConfig;
}
function mergeConfigRecursively(defaults, overrides, rootPath) {
  const merged = __spreadValues({}, defaults);
  for (const key in overrides) {
    const value = overrides[key];
    if (value == null) {
      continue;
    }
    const existing = merged[key];
    if (existing == null) {
      merged[key] = value;
      continue;
    }
    if (key === "vite" && rootPath === "") {
      merged[key] = mergeViteConfig(existing, value);
      continue;
    }
    if (Array.isArray(existing) || Array.isArray(value)) {
      merged[key] = [...arraify(existing ?? []), ...arraify(value ?? [])];
      continue;
    }
    if (isObject(existing) && isObject(value)) {
      merged[key] = mergeConfigRecursively(existing, value, rootPath ? `${rootPath}.${key}` : key);
      continue;
    }
    merged[key] = value;
  }
  return merged;
}
function mergeConfig(defaults, overrides, isRoot = true) {
  return mergeConfigRecursively(defaults, overrides, isRoot ? "" : ".");
}
export {
  AstroConfigSchema,
  LEGACY_ASTRO_CONFIG_KEYS,
  loadConfig,
  mergeConfig,
  resolveConfig,
  resolveConfigURL,
  validateConfig
};
