// astro/dist/integrations/index.js
var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
import { mergeConfig } from "../core/config.js";
import ssgAdapter from "../adapter-ssg/index.js";
async function runHookConfigSetup({
  config: _config,
  command
}) {
  if (_config.adapter) {
    _config.integrations.push(_config.adapter);
  }
  let updatedConfig = __spreadValues({}, _config);
  for (const integration of _config.integrations) {
    if (integration.hooks["astro:config:setup"]) {
      await integration.hooks["astro:config:setup"]({
        config: updatedConfig,
        command,
        addRenderer(renderer) {
          updatedConfig._ctx.renderers.push(renderer);
        },
        injectScript: (stage, content) => {
          updatedConfig._ctx.scripts.push({ stage, content });
        },
        updateConfig: (newConfig) => {
          updatedConfig = mergeConfig(updatedConfig, newConfig);
        }
      });
    }
  }
  return updatedConfig;
}
async function runHookConfigDone({ config }) {
  for (const integration of config.integrations) {
    if (integration.hooks["astro:config:done"]) {
      await integration.hooks["astro:config:done"]({
        config,
        setAdapter(adapter) {
          if (config._ctx.adapter && config._ctx.adapter.name !== adapter.name) {
            throw new Error(`Adapter already set to ${config._ctx.adapter.name}. You can only have one adapter.`);
          }
          config._ctx.adapter = adapter;
        }
      });
    }
  }
  if (!config._ctx.adapter) {
    const integration = ssgAdapter();
    config.integrations.push(integration);
    if (integration.hooks["astro:config:done"]) {
      await integration.hooks["astro:config:done"]({
        config,
        setAdapter(adapter) {
          config._ctx.adapter = adapter;
        }
      });
    }
  }
}
async function runHookServerSetup({
  config,
  server
}) {
  for (const integration of config.integrations) {
    if (integration.hooks["astro:server:setup"]) {
      await integration.hooks["astro:server:setup"]({ server });
    }
  }
}
async function runHookServerStart({
  config,
  address
}) {
  for (const integration of config.integrations) {
    if (integration.hooks["astro:server:start"]) {
      await integration.hooks["astro:server:start"]({ address });
    }
  }
}
async function runHookServerDone({ config }) {
  for (const integration of config.integrations) {
    if (integration.hooks["astro:server:done"]) {
      await integration.hooks["astro:server:done"]();
    }
  }
}
async function runHookBuildStart({
  config,
  buildConfig
}) {
  for (const integration of config.integrations) {
    if (integration.hooks["astro:build:start"]) {
      await integration.hooks["astro:build:start"]({ buildConfig });
    }
  }
}
async function runHookBuildSetup({
  config,
  vite,
  target
}) {
  for (const integration of config.integrations) {
    if (integration.hooks["astro:build:setup"]) {
      await integration.hooks["astro:build:setup"]({ vite, target });
    }
  }
}
async function runHookBuildDone({
  config,
  pages,
  routes
}) {
  for (const integration of config.integrations) {
    if (integration.hooks["astro:build:done"]) {
      await integration.hooks["astro:build:done"]({
        pages: pages.map((p) => ({ pathname: p })),
        dir: config.outDir,
        routes
      });
    }
  }
}
export {
  runHookBuildDone,
  runHookBuildSetup,
  runHookBuildStart,
  runHookConfigDone,
  runHookConfigSetup,
  runHookServerDone,
  runHookServerSetup,
  runHookServerStart
};
